main/src/ca/uwaterloo/flix/api/lsp/provider/Visitor.scala

```diff
@@ -0,0 +1,529 @@
+/*
+ * Copyright 2024 Alexander Dybdahl Troelsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider
+
+import ca.uwaterloo.flix.api.lsp.Entity
+import ca.uwaterloo.flix.language.ast.TypedAst.{
+  Root, 
+  Def, 
+  Expr, 
+  Effect, 
+  Enum,
+  Constraint, 
+  Pattern, 
+  StructField,
+  Case,
+  Instance,
+  RestrictableEnum,
+  RestrictableCase,
+  Sig,
+  Struct,
+  Trait,
+  TypeAlias,
+  TypeParam
+}
+import ca.uwaterloo.flix.language.ast.Ast.UseOrImport
+import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.Symbol
+import ca.uwaterloo.flix.api.lsp.Position
+
+/**
+  * A collection of visit functions that can apply a set of functions
+  * to an TypedAst node and all of it's children recursively.
+  */
+object Visitor {
+  /**
+    * Applies a `seenRoot` to the root AST node and recursively visits
+    * all children, applying the corresponding `seenX` functions.
+    *
+    * @param root    The AST root node.
+    * @param seenX   The `seen` function that is to be applied to X type AST nodes.
+    * @param accept  A predicate determining whether to visit a child node.
+    */
+  def visitRoot(root: Root, 
+                seenDef: Def => Unit,
+                seenEff: Effect => Unit,
+                seenEnum: Enum => Unit,
+                seenExpr: Expr => Unit,
+                seenInstance: Instance => Unit,
+                seenResEnum: RestrictableEnum => Unit,
+                seenRoot: Root => Unit,
+                seenSig: Sig => Unit,
+                seenStruct: Struct => Unit,
+                seenTrait: Trait => Unit,
+                seenTypeAlias: TypeAlias => Unit,
+                accept: SourceLocation => Boolean): Unit = {
+
+    root.defs.foreach{ case (_, defn) => {
+      val insideDefn = accept(defn.spec.loc) || accept(defn.exp.loc) || accept(defn.sym.loc)
```

Comment:
use the "clear" location `defn.loc` or `eff.loc`.

Reply:
There is the issue that `Def` do not have locations at all. Their `DefnSym`s have, the `Spec`s have and their `Expr`s have, but `Def`s in and of themselves, do not have `SourceLocation`s. Should I use the symbols locations or something?

Reply:
Yeah good point. I also understand your point about the locations not covering the entire "thing" defs/effect/etc. Maybe we should have some extra PRs to add locations to defs fx.

Reply:
Yeah. It's an issue I've run into a couple of times. How should I handle these situations for now?